mne/minimum_norm/spatial_resolution.py

```diff
@@ -0,0 +1,429 @@
+    src : instance of SourceSpaces
+        Source space object from forward or inverse operator.
+    function : 'psf' | 'ctf'
+        Whether to compute metrics for columns (point-spread functions, PSFs)
+        or rows (cross-talk functions, CTFs) of the resolution matrix.
+    kind : str
+        What kind of resolution feature to consider.
+        Options are 'localization_error', 'spatial_extent', and 'amplitude'.
+    metric : str
+        The exact resolution metric for the kind of resolution feature.
+        This must match with 'kind'. The allowed options are:
+
+        - ``kind='localization_error'``:
+
+          - ``metric='peak'``
+              Peak localization error (PLE), Euclidean distance between
+              peak and true source location.
+          - ``metric='cog'``
+              Centre-of-gravity localisation error (CoG), Euclidean
+              distance between CoG and true source location.
+
+        - ``kind='spatial_extent'``:
+
+          - ``metric='sd'``
+              spatial deviation (e.g. [1,2]_).
+          - ``metric='maxrad'``
+              maximum radius to 50% of max amplitude.
+        - ``kind='relative_amplitude'``:
+
+          - ``metric='peak'``
+              Ratio between absolute maximum amplitudes of peaks per
+              location and maximum peak across locations.
+          - ``metric='sum'``
+              Ratio between sums of absolute amplitudes.
+
+    threshold : float
+        Amplitude fraction threshold for spatial extent metric 'maxrad'.
+        Defaults to 0.5.
+
+    Returns
+    -------
+    resolution_metric : instance of SourceEstimate
+        The resolution metric.
+
+    Notes
+    -----
+    For details, see [1]_ [2]_.
+
+    .. versionadded:: 0.20
+
+    References
+    ----------
+    .. [1] Molins A, Stufflebeam S M, Brown E N, Hämäläinen M S (2008).
+           Quantification of the benefit from integrating MEG and EEG data in
+           minimum l2-norm estimation. Neuroimage, 42(3):1069-77.
+    .. [2] Hauk O, Stenroos M, Treder M (2019). "Towards an Objective
+           Evaluation of EEG/MEG Source Estimation Methods: The Linear Tool
+           Kit", bioRxiv, doi: https://doi.org/10.1101/672956.
+    """
+    # Check if input options are valid
+    if kind == 'localization_error':
+        if metric not in ['peak', 'cog']:
+            raise ValueError('Not an allowed metric for localization_error: %s'
+                             % metric)
+
+    elif kind == 'spatial_extent':
+        if metric not in ['sd', 'maxrad']:
+            raise ValueError('Not an allowed metric for spatial_extent: %s'
+                             % metric)
+
+    elif kind == 'relative_amplitude':
+        if metric not in ['peak', 'sum']:
+            raise ValueError('Not an allowed metric for amplitude: %s'
+                             % metric)
+
+    else:
+        raise ValueError('Not a recognised kind of resolution feature: %s.'
+                         % kind)
```

Comment:
If we collapse `kind` and `metric` these 18 lines get simplified to 3 lines (with minimal disambiguation added for the two meanings of `peak`):

```
metrics = ('peak_err', 'cog', 'sd', 'maxrad', 'peak_ratio', 'sum')
if metric not in metrics:
    raise ValueError('"%s" is not a recognized metric.' % metric)
```

Later lines of code will need to change too, but not a lot and not in a way that makes the code harder to read/maintain, as far as I can tell.

So I'm still really not convinced that keeping the arguments separate is the right way to "force users to think about the different aspects of resolution (e.g. localisation, spatial extent, amplitude estimation)", especially at the expense of having more lines of code to maintain (lines that contain, in my opinion, unecessarily nested logic).  A well-written docstring, well-chosen parameter values, and thorough examples/tutorials seems to me to be at least as effective (if not moreso) at achieving that goal.

If you still think that mixing `kind` and `metric` into a single string (e.g., `spatial_deviation`, `spatial_max_rad`, `spatial_area`, `loc_error_peak`, `loc_error_cog`, ...) is insufficient, then I think making the user choose among 3 different public *functions* for the different kinds of metric makes more sense to me than the current implementation.